### PR TITLE
Fix invalid pointer errors in Burnout 2.

### PR DESCRIPTION
**Pull request by phire**
Yet another story of games loading weird shit into registers.

For some reason, Burnout 2 would (in rare situations) load invalid addresses into cp_state.array_bases. What would the real hardware do in this situation? Who knows, Burnout 2 doesn't actually enable the vertex array with the invalid address so nothing kinky happens.

But dolphin tries to optimise things and starts using the address as soon as it is loaded into memory. This causes GetPointer (which is now much more vocal) to throw an error.

The Fix:  We don't call GetPointer until we are sure the vertex array has been enabled.

### CODE REVIEW

File: Source/Core/VideoCommon/CPMemory.h [hunk 1]
```diff
@@ -88,6 +88,12 @@ union TVtxDesc
 	{
 		u32 Hex0, Hex1;
 	};
+
+	// Easily index into the Position..Tex7Coord fields.
+	u32 getVertexArrayStatus(int idx)
```
Comment (lioncash) [inline Source/Core/VideoCommon/CPMemory.h]: This needs to be uppercased.

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 2]
```diff
@@ -263,10 +265,15 @@ void FillCPMemoryArray(u32 *memory)
 	}
 }
 
-void RecomputeCachedArraybases()
+void ComputeCachedArraybases()
```
Comment (lioncash) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: `bases` should be uppercased here.

**Comment by degasus** [comment]
Both commits looks good to me, but I fear this will call GetPointer too often.
What do you think about adding a dirty flag which is set on writes to vtx_desc and array_bases?
RefreshPointer is called quite often.

**Comment by phire** [comment]
GetPointer is reasonably simple these days, it only handles the ranges of 0x00000000 to 0x01800000 and 0x10000000 to 0x14000000 (if it's a Wii) 

Additionally, I really suspect that it will be rare for an array to be used more than once. It will only happen if the game is "instancing". Most of the time all the arrays will be updated every drawcall or unused.

File: Source/Core/VideoCommon/VertexLoaderManager.h [hunk 3]
```diff
@@ -24,6 +24,9 @@ namespace VertexLoaderManager
 	void AppendListToString(std::string *dest);
 
 	NativeVertexFormat* GetCurrentVertexFormat();
+
+	// Resloved pointers to array bases. Used by vertex loaders.
```
Comment (BhaaLseN) [inline Source/Core/VideoCommon/VertexLoaderManager.h]: Typo :)

**Comment by phire** [comment]
Adding a small dirty flag optimisation brings the performance of vertex loader heavy games up to about the same as master.

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 4]
```diff
@@ -232,8 +259,7 @@ void LoadCPReg(u32 sub_cmd, u32 value, bool is_preprocess)
 	// Pointers to vertex arrays in GC RAM
 	case 0xA0:
 		state->array_bases[sub_cmd & 0xF] = value;
-		if (update_global_state)
-			cached_arraybases[sub_cmd & 0xF] = Memory::GetPointer(value);
+		state->bases_dirty = true;
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: also on writes to vtx_desc

File: Source/Core/VideoCommon/VertexLoaderManager.cpp [hunk 5]
```diff
@@ -53,6 +54,27 @@ void Shutdown()
 	s_native_vertex_map.clear();
 }
 
+void UpdateVertexArrayPointers()
+{
+	// Anything to update?
+	if (!g_main_cp_state.bases_dirty)
```
Comment (degasus) [inline Source/Core/VideoCommon/VertexLoaderManager.cpp]: I'd move this to the caller, so you don't have to care about videosw

**Comment by degasus** [comment]
LGTM

**Comment by JMC47** [comment]
Performance is fine, LGTM

**Comment by Tilka** [comment]
Oh well.

**Comment by degasus** [comment]
Sorry @Tilka, I haven't seen that you are reviewing this PR right now :/

**Pull request merged**